### ISSUE TITLE
API announces to passive neighbors

### ISSUE THREAD

**Issue by ykondratov** [Exa-Networks/exabgp#926]
##### Summary

Maybe not an issue, but a question. I run exabgp with config for passive neighbors from a subnet. Peering is established successfully, but when I try to announce routes via API, nothing is actually sent, even though I see that command is accepted. Maybe I need to turn on some option in env file or config in order to send updates to passive neighbors?

##### OS

```
Linux
```

```
Linux exa 4.19.78-0-vanilla #1-Alpine SMP Thu Oct 10 15:06:04 UTC 2019 x86_64 Linux
```

##### Version

```
ExaBGP : 4.1.2-d006a34a
Python : 3.7.5 (default, Oct 16 2019, 06:53:15)  [GCC 8.3.0]
Uname  : Linux exa 4.19.78-0-vanilla #1-Alpine SMP Thu Oct 10 15:06:04 UTC 2019 x86_64
Root   : /usr

```

##### Installation

```
pip
```

##### Environment 

```
[exabgp.api]
ack = false

[exabgp.daemon]
user = 'root'

[exabgp.log]
all = true
packets = true

[exabgp.tcp]
bind = '10.0.3.164'

```

##### Configuration

```
template {
    neighbor controller {
        family {
            ipv4 unicast;
        }
    }
}



#neighbor range
neighbor 10.0.3.0/24 {
    inherit controller;
    passive true;
    router-id 10.0.3.164;
    local-address 10.0.3.164;
    local-as 65010;
    peer-as 65010;
}

```

##### Program output

```
exabgp -d /etc/exabgp/exabgp.conf
14:08:03 | 6672   | welcome       | Thank you for using ExaBGP
14:08:03 | 6672   | version       | 4.1.2-d006a34a
14:08:03 | 6672   | interpreter   | 3.7.5 (default, Oct 16 2019, 06:53:15)  [GCC 8.3.0]
14:08:03 | 6672   | os            | Linux exa 4.19.78-0-vanilla #1-Alpine SMP Thu Oct 10 15:06:04 UTC 2019 x86_64
14:08:03 | 6672   | installation  | /usr
14:08:03 | 6672   | cli control   | named pipes for the cli are:
14:08:03 | 6672   | cli control   | to send commands  /run/exabgp.in
14:08:03 | 6672   | cli control   | to read responses /run/exabgp.out
14:08:03 | 6672   | network       | listening for BGP session(s) on 10.0.3.164:179
14:08:03 | 6672   | configuration | performing reload of exabgp 4.1.2-d006a34a
14:08:03 | 6672   | configuration | > template         |
14:08:03 | 6672   | configuration | > neighbor         | 'controller'
14:08:03 | 6672   | configuration | > family           |
14:08:03 | 6672   | configuration | . ipv4             | 'unicast'
14:08:03 | 6672   | configuration | < family           |
14:08:03 | 6672   | configuration | < neighbor         |
14:08:03 | 6672   | configuration | < template         |
14:08:03 | 6672   | configuration | > neighbor         | '10.0.3.0/24'
14:08:03 | 6672   | configuration | . inherit          | 'controller'
14:08:03 | 6672   | configuration | . passive          | 'true'
14:08:03 | 6672   | configuration | . router-id        | '10.0.3.164'
14:08:03 | 6672   | configuration | . local-address    | '10.0.3.164'
14:08:03 | 6672   | configuration | . local-as         | '65010'
14:08:03 | 6672   | configuration | . peer-as          | '65010'
14:08:03 | 6672   | configuration | < neighbor         |
14:08:03 | 6672   | reactor       | new peer: neighbor 10.0.3.0/24 local-ip 10.0.3.164 local-as 65010 peer-as 65010 router-id 10.0.3.164 family-allowed in-open
14:08:03 | 6672   | network       | listening for BGP session(s) on 10.0.3.164:179
14:08:03 | 6672   | reactor       | loaded new configuration successfully
14:08:03 | 6672   | process       | forked process api-internal-cli-5e31d910
14:08:22 | 6672   | reactor       | async | 696e5b46-f01e-11e9-86db-5254004584f5 | checking for new connection(s)
14:08:22 | 6672   | network       | connection from 10.0.3.164
14:08:22 | 6672   | network       | new connection received incoming-1 10.0.3.132-10.0.3.164
14:08:22 | 6672   | peer-1        | state machine for the peer is IDLE
14:08:22 | 6672   | reactor       | initialising connection to peer-1
14:08:22 | 6672   | incoming-1    | sending TCP payload (  49) FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF 0031 0104 FDF2 00B4 0A00 03A4 1402 0601 0400 0100 0102 0641 0400 00FD F202 0206 00
14:08:22 | 6672   | incoming-1    | >> OPEN version=4 asn=65010 hold_time=180 router_id=10.0.3.164 capabilities=[Multiprotocol(ipv4 unicast), Extended Message(65535), ASN4(65010)]
14:08:22 | 6672   | incoming-1    | received TCP payload (  19) FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF 0055 01
14:08:22 | 6672   | incoming-1    | received TCP payload (  66) 04FD F200 B4AC 1200 1238 0206 0104 0001 0001 0202 8000 0202 0200 0206 4104 0000 FDF2 0206 4504 0001 0101 0210 490E 0C37 6363 3261 3430 6161 3764 6100 0204 4002 0078
14:08:22 | 6672   | incoming-1    | << message of type OPEN
14:08:22 | 6672   | incoming-1    | << OPEN version=4 asn=65010 hold_time=180 router_id=172.18.0.18 capabilities=[Multiprotocol(ipv4 unicast), Route Refresh, Graceful Restart Flags 0x0 Time 120 , ASN4(65010), AddPath(receive ipv4 unicast), Unassigned 73, Route Refresh]
14:08:22 | 6672   | ka-incoming-1 | receive-timer 60 second(s) left
14:08:22 | 6672   | incoming-1    | sending TCP payload (  19) FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF 0013 04
14:08:22 | 6672   | incoming-1    | >> KEEPALIVE (OPENCONFIRM)
14:08:22 | 6672   | incoming-1    | received TCP payload (  19) FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF 0013 04
14:08:22 | 6672   | incoming-1    | << message of type KEEPALIVE
14:08:22 | 6672   | ka-incoming-1 | receive-timer 180 second(s) left
14:08:22 | 6672   | reactor       | connected to peer-1 with incoming-1 10.0.3.132-10.0.3.164
14:08:22 | 6672   | incoming-1    | sending TCP payload (  23) FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF 0017 0200 0000 00
14:08:22 | 6672   | incoming-1    | >> EOR ipv4 unicast
14:08:22 | 6672   | peer-1        | >> EOR(s)
14:08:22 | 6672   | reactor       | async | 69a2c192-f01e-11e9-86db-5254004584f5 | checking for new connection(s)
14:08:22 | 6672   | network       | connection from 10.0.3.164
14:08:22 | 6672   | network       | new connection received incoming-2 10.0.3.128-10.0.3.164
14:08:22 | 6672   | peer-1        | state machine for the peer is IDLE
14:08:22 | 6672   | reactor       | initialising connection to peer-1
14:08:22 | 6672   | incoming-2    | sending TCP payload (  49) FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF 0031 0104 FDF2 00B4 0A00 03A4 1402 0601 0400 0100 0102 0641 0400 00FD F202 0206 00
14:08:22 | 6672   | incoming-2    | >> OPEN version=4 asn=65010 hold_time=180 router_id=10.0.3.164 capabilities=[Multiprotocol(ipv4 unicast), Extended Message(65535), ASN4(65010)]
14:08:22 | 6672   | incoming-2    | received TCP payload (  19) FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF 0055 01
14:08:22 | 6672   | incoming-2    | received TCP payload (  66) 04FD F200 B4AC 1200 0838 0206 0104 0001 0001 0202 8000 0202 0200 0206 4104 0000 FDF2 0206 4504 0001 0101 0210 490E 0C34 3463 3761 3238 3239 6461 6600 0204 4002 0078
14:08:22 | 6672   | incoming-2    | << message of type OPEN
14:08:22 | 6672   | incoming-2    | << OPEN version=4 asn=65010 hold_time=180 router_id=172.18.0.8 capabilities=[Multiprotocol(ipv4 unicast), Route Refresh, Graceful Restart Flags 0x0 Time 120 , ASN4(65010), AddPath(receive ipv4 unicast), Unassigned 73, Route Refresh]
14:08:22 | 6672   | ka-incoming-2 | receive-timer 60 second(s) left
14:08:22 | 6672   | incoming-2    | sending TCP payload (  19) FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF 0013 04
14:08:22 | 6672   | incoming-2    | >> KEEPALIVE (OPENCONFIRM)
14:08:22 | 6672   | incoming-2    | received TCP payload (  19) FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF 0013 04
14:08:22 | 6672   | incoming-2    | << message of type KEEPALIVE
14:08:22 | 6672   | ka-incoming-2 | receive-timer 180 second(s) left
14:08:22 | 6672   | reactor       | connected to peer-1 with incoming-2 10.0.3.128-10.0.3.164
14:08:22 | 6672   | incoming-2    | sending TCP payload (  23) FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF 0017 0200 0000 00
14:08:22 | 6672   | incoming-2    | >> EOR ipv4 unicast
14:08:22 | 6672   | peer-1        | >> EOR(s)
14:08:23 | 6672   | ka-incoming-2 | receive-timer 179 second(s) left
14:08:23 | 6672   | ka-incoming-2 | send-timer 59 second(s) left
14:08:23 | 6672   | ka-incoming-1 | receive-timer 179 second(s) left
14:08:23 | 6672   | ka-incoming-1 | send-timer 59 second(s) left
14:08:23 | 6672   | incoming-1    | received TCP payload (  19) FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF 0034 02
14:08:23 | 6672   | incoming-1    | received TCP payload (  33) 0000 001C 4001 0100 5002 0006 0201 0000 FDEA 4003 040A 1E1E 0240 0504 0000 0064 00
14:08:23 | 6672   | incoming-1    | << message of type UPDATE
14:08:23 | 6672   | parser        | parsing UPDATE (  33) 0000 001C 4001 0100 5002 0006 0201 0000 FDEA 4003 040A 1E1E 0240 0504 0000 0064 00
14:08:23 | 6672   | routes        | withdrawn NLRI none
14:08:23 | 6672   | parser        | attribute origin             flag 0x40 type 0x01 len 0x01 payload 00
14:08:23 | 6672   | parser        | attribute as-path            flag 0x50 type 0x02 len 0x06 payload 0201 0000 FDEA
14:08:23 | 6672   | parser        | attribute next-hop           flag 0x40 type 0x03 len 0x04 payload 0A1E 1E02
14:08:23 | 6672   | parser        | attribute local-preference   flag 0x40 type 0x05 len 0x04 payload 0000 0064
14:08:23 | 6672   | parser        | NLRI      ipv4 unicast       without path-information     payload 00
14:08:23 | 6672   | routes        | announced NLRI 0.0.0.0/0 next-hop 10.30.30.2
14:08:23 | 6672   | peer-1        | << UPDATE #1
14:08:23 | 6672   | peer-1        |    UPDATE #1 nlri  (   1) 0.0.0.0/0 next-hop 10.30.30.2
14:08:23 | 6672   | incoming-2    | received TCP payload (  19) FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF 0039 02
14:08:23 | 6672   | incoming-2    | received TCP payload (  38) 0000 001D 4001 0102 5002 0000 4003 04AC 1200 0880 0404 0000 0000 4005 0400 0000 6420 0101 0101
14:08:23 | 6672   | incoming-2    | << message of type UPDATE
14:08:23 | 6672   | parser        | parsing UPDATE (  38) 0000 001D 4001 0102 5002 0000 4003 04AC 1200 0880 0404 0000 0000 4005 0400 0000 6420 0101 0101
14:08:23 | 6672   | routes        | withdrawn NLRI none
14:08:23 | 6672   | parser        | attribute origin             flag 0x40 type 0x01 len 0x01 payload 02
14:08:23 | 6672   | parser        | attribute as-path            flag 0x50 type 0x02 len 0x00
14:08:23 | 6672   | parser        | attribute next-hop           flag 0x40 type 0x03 len 0x04 payload AC12 0008
14:08:23 | 6672   | parser        | attribute med                flag 0x80 type 0x04 len 0x04 payload 0000 0000
14:08:23 | 6672   | parser        | attribute local-preference   flag 0x40 type 0x05 len 0x04 payload 0000 0064
14:08:23 | 6672   | parser        | NLRI      ipv4 unicast       without path-information     payload 2001 0101 01
14:08:23 | 6672   | routes        | announced NLRI 1.1.1.1/32 next-hop 172.18.0.8
14:08:23 | 6672   | peer-1        | << UPDATE #1
14:08:23 | 6672   | peer-1        |    UPDATE #1 nlri  (   5) 1.1.1.1/32 next-hop 172.18.0.8
14:08:23 | 6672   | incoming-2    | received TCP payload (  19) FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF 0037 02
14:08:23 | 6672   | incoming-2    | received TCP payload (  36) 0000 001D 4001 0102 5002 0000 4003 04AC 1200 0880 0404 0000 0000 4005 0400 0000 6410 AC12
14:08:23 | 6672   | incoming-2    | << message of type UPDATE
14:08:23 | 6672   | parser        | parsing UPDATE (  36) 0000 001D 4001 0102 5002 0000 4003 04AC 1200 0880 0404 0000 0000 4005 0400 0000 6410 AC12
14:08:23 | 6672   | routes        | withdrawn NLRI none
14:08:23 | 6672   | parser        | NLRI      ipv4 unicast       without path-information     payload 10AC 12
14:08:23 | 6672   | routes        | announced NLRI 172.18.0.0/16 next-hop 172.18.0.8
14:08:23 | 6672   | peer-1        | << UPDATE #2
14:08:23 | 6672   | peer-1        |    UPDATE #2 nlri  (   3) 172.18.0.0/16 next-hop 172.18.0.8
14:08:24 | 6672   | ka-incoming-2 | receive-timer 179 second(s) left
14:08:24 | 6672   | ka-incoming-2 | send-timer 58 second(s) left
14:09:36 | 6672   | process       | command from process api-internal-cli-5e31d910 : neighbor 10.0.3.132 announce route 1.1.1.1/32 next-hop 10.0.3.164 as-path [ ] local-preference 150
14:09:36 | 6672   | reactor       | async | api-internal-cli-5e31d910 | neighbor 10.0.3.132 announce route 1.1.1.1/32 next-hop 10.0.3.164 as-path [ ] local-preference 150
14:09:36 | 6672   | configuration | . route            | '1.1.1.1/32' 'next-hop' '10.0.3.164' 'as-path' '[' ']' 'local-preference' '150'
14:09:36 | 6672   | api           | route added to neighbor 10.0.3.132 local-ip 10.0.3.164 local-as 65010 peer-as 65010 router-id 10.0.3.164 family-allowed in-open : 1.1.1.1/32 next-hop 10.0.3.164 as-path [ ] local-preference 150
14:09:37 | 6672   | ka-incoming-2 | receive-timer 165 second(s) left
14:09:37 | 6672   | ka-incoming-2 | send-timer 45 second(s) left
```

##### Steps to reproduce

Run exabgp with config for passive neighbor, wait until peering is established, send route announce via exabgpcli. Command is accepted, but no update actually sent (I've checked with tcpdump). 

If I change conf file with explicit neighbors (not passive), then everything works fine. 
```
 exabgpcli nei 10.0.3.132 announce route 1.1.1.1/32 next-hop 10.0.3.164 as-path [ ] local-preference 150
command: neighbor 10.0.3.132 announce route 1.1.1.1/32 next-hop 10.0.3.164 as-path [ ] local-preference 150

warning: no end of command message received
warning: normal if exabgp.api.ack is set to false otherwise some data may get stuck on the pipe
warning: otherwise it may cause exabgp reactor to block

```

##### Importance

Not in production. 
```

```

**Comment by ykondratov** [comment]
I try to send it via CLI like this: 

```
exabgpcli nei 10.0.3.132 announce route 1.1.1.1/32 next-hop 10.0.3.164 as-path [ ] local-preference 150
```

Still no update sent to neighbor.

**Comment by thomas-mangin** [comment]
Could you try without filtering with "nei .." to see what happens with "announce route .." please?

**Comment by ykondratov** [comment]
I tried to announce without "nei" keyword: 
```
exabgpcli announce route 1.1.1.1/32 next-hop 10.0.3.164 as-path [ ] local-preference 150
```
and in packet capture on the outgoing interface I can see one BGP Update packet, which is sent to the first established neighbor (the one with lowest RID)
![image](https://user-images.githubusercontent.com/41232010/67268348-e1c00e80-f4bc-11e9-9068-94cb4c486ffb.png)


In the debug of exabgp I can see that routes added to all neighbors. But in fact only one update is actually sent. 
```
08:00:24 | 3151   | ka-incoming-1 | send-timer 54 second(s) left
08:00:24 | 3151   | ka-incoming-2 | receive-timer 128 second(s) left
08:00:24 | 3151   | ka-incoming-2 | send-timer 8 second(s) left
08:00:24 | 3151   | process       | command from process api-internal-cli-90585146 : announce route 1.1.1.1/32 next-hop 10.0.3.164 as-path [ ] local-preference 150
08:00:24 | 3151   | reactor       | async | api-internal-cli-90585146 | announce route 1.1.1.1/32 next-hop 10.0.3.164 as-path [ ] local-preference 150
08:00:24 | 3151   | configuration | . route            | '1.1.1.1/32' 'next-hop' '10.0.3.164' 'as-path' '[' ']' 'local-preference' '150'
08:00:24 | 3151   | api           | route added to neighbor 10.0.3.0/24 local-ip 10.0.3.164 local-as 65010 peer-as 65010 router-id 10.0.3.164 family-allowed in-open, neighbor 10.0.3.128 local-ip 10.0.3.164 local-as 65010 peer-as 65010 router-id 10.0.3.164 family-allowed in-open, neighbor 10.0.3.132 local-ip 10.0.3.164 local-as 65010 peer-as 65010 router-id 10.0.3.164 family-allowed in-open : 1.1.1.1/32 next-hop 10.0.3.164 as-path [ ] local-preference 150
08:00:24 | 3151   | incoming-1    | sending TCP payload (  49) FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF 0031 0200 0000 1540 0101 0040 0200 4003 040A 0003 A440 0504 0000 0096 2001 0101 01
08:00:24 | 3151   | incoming-1    | >> 1 UPDATE(s)
08:00:25 | 3151   | ka-incoming-1 | receive-timer 173 second(s) left
08:00:25 | 3151   | ka-incoming-1 | send-timer 53 second(s) left
08:00:25 | 3151   | ka-incoming-2 | receive-timer 127 second(s) left

```

**Comment by thomas-mangin** [comment]
Not many people would use passive. If the other neighbors have not connected it is right for the route to not be sent to them. Also if they connect later, I am not sure if they will get it or not once th API call was done. They should.

**Comment by ykondratov** [comment]
I use passive to workaround this issue https://github.com/Exa-Networks/exabgp/issues/836
So that when a new peer is added I don't have to reread config and restart the api process. 

So it is a feature-request I guess: to be able to send announce/withdraw to passive neighbors per neighbor?

**Comment by thomas-mangin** [comment]
If I cal recall the code correctly, the configuration is parsed and when as "/net" is used, the configuration is duplicated and the IP updated. So if the ADJ-RIB-OUT is enabled (caching of route sent), then all sent/received routes should be in the per peer rib and be sent/resent on reconnection.

**Comment by thomas-mangin** [comment]
```
warning: no end of command message received
warning: normal if exabgp.api.ack is set to false otherwise some data may get stuck on the pipe
warning: otherwise it may cause exabgp reactor to block
```

Are you still seing this -- you should not !

**Comment by ykondratov** [comment]
I've checked once more. Maybe I am doing smth wrong, but as far as I can see, BGP update is not sent to all neighbors in Established state. My lab is with 2 neighbors. 
```
exavm:~# exabgpcli show nei sum ext
command: show neighbor summary extensive
Peer            AS        up/down state       |     #sent     #recvd
10.0.3.0/24     65010        down idle                  0          0
10.0.3.132      65010     0:02:33 established           1          1
10.0.3.128      65010     0:01:38 established           1          2

```
I use CLI like this:
```
exavm:~# exabgpcli announce route 1.1.1.5/32 next-hop 10.0.3.164 as-path [ ] local-preference 150

warning: no end of command message received
warning: normal if exabgp.api.ack is set to false otherwise some data may get stuck on the pipe
warning: otherwise it may cause exabgp reactor to block

```
And the update is sent only to 10.0.3.132. I guess it came up first, because earlier update was sent only to 10.0.3.128. So in my setup BGP update is sent to neighbor, who established the peering first. In RIB-OUT prefixes are present:
```
exavm:~# exabgpcli show adj-rib out
neighbor 10.0.3.0/24 ipv4 unicast 1.1.1.5/32 next-hop 10.0.3.164
neighbor 10.0.3.0/24 ipv4 unicast 1.1.1.6/32 next-hop 10.0.3.164
neighbor 10.0.3.0/24 ipv4 unicast 1.1.1.7/32 next-hop 10.0.3.164
neighbor 10.0.3.132 ipv4 unicast 1.1.1.5/32 next-hop 10.0.3.164
neighbor 10.0.3.132 ipv4 unicast 1.1.1.6/32 next-hop 10.0.3.164
neighbor 10.0.3.132 ipv4 unicast 1.1.1.7/32 next-hop 10.0.3.164
neighbor 10.0.3.128 ipv4 unicast 1.1.1.5/32 next-hop 10.0.3.164
neighbor 10.0.3.128 ipv4 unicast 1.1.1.6/32 next-hop 10.0.3.164
neighbor 10.0.3.128 ipv4 unicast 1.1.1.7/32 next-hop 10.0.3.164

```

I've turned off acks for messages, so I keep seeing warnings, but they explicitly state, that this is normal with acks disabled. This has changed? My version is:
```
ExaBGP : 4.1.2-d006a34a
Python : 3.7.5 (default, Oct 16 2019, 06:53:15)  [GCC 8.3.0]
Uname  : Linux rvnfd 4.19.78-0-vanilla #1-Alpine SMP Thu Oct 10 15:06:04 UTC 2019 x86_64
Root   : /usr
```

Summing up the overall topic: I cannot send per neighbor update if neighbor came up as passive? 
If so, I think I can workround this, but just wanted to ask if this behaviour cannot be changed via some option in env file.

**Comment by thomas-mangin** [comment]
sorry for the delay, this should do the trick !
